crates/metadata-store/src/omnipaxos/store.rs

```diff
@@ -0,0 +1,1232 @@
+use restate_types::{GenerationalNodeId, PlainNodeId, Version};
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::future::Future;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::{mpsc, oneshot};
+use tokio::time;
+use tokio::time::MissedTickBehavior;
+use tracing::{debug, info, instrument, trace, warn};
+use ulid::Ulid;
+
+type OmniPaxos = omnipaxos::OmniPaxos<Request, RocksDbStorage<Request>>;
+
+#[derive(Debug, thiserror::Error)]
+#[error("failed accessing storage: {0}")]
+pub struct StorageError(String);
+
+impl From<Box<dyn std::error::Error>> for StorageError {
+    fn from(value: Box<dyn std::error::Error>) -> Self {
+        StorageError(value.to_string())
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+enum DecidedEntriesError {
+    #[error(transparent)]
+    Storage(#[from] StorageError),
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("invalid nodes configuration: {0}")]
+pub struct InvalidConfiguration(String);
+
+#[derive(Debug, thiserror::Error)]
+pub enum InitializeMetadataStoreError {
+    #[error(transparent)]
+    InvalidConfiguration(#[from] InvalidConfiguration),
+    #[error("failed encoding nodes configuration: {0}")]
+    Codec(#[from] StorageEncodeError),
+    #[error("failed initializing the storage: {0}")]
+    Storage(#[from] StorageError),
+}
+
+impl Entry for Request {
+    type Snapshot = NoSnapshot;
+}
+
+pub struct OmniPaxosMetadataStore {
+    connection_manager: Arc<ArcSwapOption<ConnectionManager<OmniPaxosMessage>>>,
+    rocksdb_storage: RocksDbStorage<Request>,
+    storage_id: StorageId,
+    metadata_writer: Option<MetadataWriter>,
+
+    request_tx: RequestSender,
+    request_rx: RequestReceiver,
+
+    provision_tx: ProvisionSender,
+    provision_rx: Option<ProvisionReceiver>,
+
+    join_cluster_tx: JoinClusterSender,
+    join_cluster_rx: JoinClusterReceiver,
+}
+
+impl OmniPaxosMetadataStore {
+    pub async fn create(
+        rocks_db_options: BoxedLiveLoad<RocksDbOptions>,
+        metadata_writer: Option<MetadataWriter>,
+    ) -> Result<Self, BuildError> {
+        let (request_tx, request_rx) = mpsc::channel(2);
+        let (provision_tx, provision_rx) = mpsc::channel(1);
+        let (join_cluster_tx, join_cluster_rx) = mpsc::channel(1);
+
+        let rocksdb_storage =
+            RocksDbStorage::create(&Configuration::pinned().metadata_store, rocks_db_options)
+                .await?;
+
+        // make sure that the storage is initialized with a storage id to be able to detect disk losses
+        let storage_id = if let Some(storage_id) = rocksdb_storage
+            .get_storage_id()
+            .map_err(|err| BuildError::InitStorage(err.to_string()))?
+        {
+            storage_id
+        } else {
+            let storage_id = random();
+            rocksdb_storage
+                .set_storage_id(storage_id)
+                .map_err(|err| BuildError::InitStorage(err.to_string()))?;
+            storage_id
+        };
```

Comment:
The durable storage_id together with the node's `PlainNodeId` constitute the `MemberId` which is used to check whether a node is part of a configuration or not. In case of a disk loss we will generate a new `StorageId` which requires a reconfiguration for a node to rejoin an omni paxos cluster.